starlette/websocket.py

```diff
@@ -0,0 +1,235 @@
+        receives a binary message).
+        """
+        return 1003
+
+    @property
+    def WS_1007_INALID_DATA(self):
+        """
+        1007 indicates that an endpoint is terminating the connection
+        because it has received data within a message that was not
+        consistent with the type of the message (e.g., non-UTF-8 [RFC3629]
+        data within a text message).
+        """
+        return 1007
+
+    @property
+    def WS_1008_POLICY_VIOLATION(self):
+        """
+        1008 indicates that an endpoint is terminating the connection
+        because it has received a message that violates its policy.  This
+        is a generic status code that can be returned when there is no
+        other more suitable status code (e.g., 1003 or 1009) or if there
+        is a need to hide specific details about the policy.
+        """
+        return 1008
+
+    @property
+    def WS_1009_TOO_BIG(self):
+        """
+        1009 indicates that an endpoint is terminating the connection
+        because it has received a message that is too big for it to
+        process.
+        """
+        return 1009
+
+    @property
+    def WS_1010_TLS_FAIL(self):
+        """
+        1010 indicates that an endpoint (client) is terminating the
+        connection because it has expected the server to negotiate one or
+        more extension, but the server didn't return them in the response
+        message of the WebSocket handshake.  The list of extensions that
+        """
+        return 1010
+
+
+status = Status()
+
+
+class WSState(enum.Enum):
+    CONNECTING = 0
+    CONNECTED = 1
+    CLOSED = 2
+
+
+class WebSocket(object):
+    """
+    https://github.com/django/asgiref/blob/master/specs/www.rst
+    """
+
+    def __init__(
+        self, scope: dict, asgi_receive: typing.Callable, asgi_send: typing.Callable
+    ) -> None:
+
+        if scope.get("type") != "websocket":
+            raise WebSocketProtocolError(detail="Not a websocket scope")
+
+        self.request = Request(scope)
+        self._asgi_send = asgi_send
+        self._asgi_receive = asgi_receive
+        self._state = WSState.CLOSED
+
+    @property
+    def subprotocols(self) -> list:
+        return self.request.get("subprotocols", [])
+
+    @property
+    def connected(self):
+        return self._state == WSState.CONNECTED
+
+    @property
+    def connecting(self):
+        return self._state == WSState.CONNECTING
+
+    @property
+    def closed(self):
+        return self._state == WSState.CLOSED
+
+    async def connect(
+        self,
+        subprotocol: str = None,
+        close: bool = False,
+        close_code: int = status.WS_1000_OK,
+    ) -> None:
+        # Accept or Refuse an incoming connection
+        if self._state != WSState.CLOSED:
+            # Try to send a close and be friendly to the otherside before raising
+            await self.close(code=status.WS_1001_LEAVING)
+
+            raise WebSocketProtocolError(
```

Comment:
I've got a bunch of similar things where it currently just uses `assert` to ensure state transitions, that would be far better as `WebSocketProtocolError` exceptions. I partly didn't get around to that yet because it made getting the 100% coverage a bit more work. Any PRs around that would be very gratefully accepted.

It's worth taking a look at the state transitions in `websockets.py` - you'll see that there's a client state and an application state, and that we perform the transitions and checks in `.send()` and `.receive()`.